Refactor AffExpr and QuadExpr to use OrderedDict for internal storage.

mlubin opened the pull request:
This means that duplicates are no longer possible. I left zeros in there for now. We might decide to always remove them. I also expect some bikeshedding on the API (`push!`/`append!` are weird names now).

See https://github.com/JuliaOpt/MathOptInterface.jl/issues/234#issuecomment-366868878 for initial benchmarks. I haven't benchmarked the new setup but paid attention to implementing things efficiently. It's not obvious how to do a fair comparison given that `master` doesn't currently remove duplicates, and too much has changed since 0.18.

We can probably close https://github.com/JuliaOpt/MathOptInterface.jl/issues/234 by saying that solvers don't need to accept duplicates.

I'll track down any coverage gaps if the diff is not 100% covered.

@rdeits @vtjeng @ExpandingMan

vtjeng commented:
Thanks @mlubin! (For the problems that I've been working with, simplifying the `AffExpr`s has in some instances been taking the majority of the time, so this would be a great help).

@rdeits - this means that simplification will no longer be needed in the next release of `ConditionalJuMP` that builds on this version of `JuMP`, right?

rdeits commented:
Yup, although that might just happen when I rewrite it for MathOptInterface.

Anyway, this is great!

blegat commented on src/affexpr.jl:
```diff
@@ -15,23 +15,83 @@
+# Utilities for OrderedDict
+function add_or_set!(dict::OrderedDict{K,V}, k::K, v::V) where {K,V}
+    # TODO: This unnecessarily requires two lookups for k.
+    # TODO: Decide if we want to drop zeros here.
```
I am in favor of removing the zeros here

mlubin replied:
This could have performance implications that I don't understand well. I'd want to check that the implementation handles deletion and re-adding efficiently or that benchmarks don't show a regression if we delete zeros.

blegat commented on src/affexpr.jl:
```diff
@@ -15,23 +15,83 @@
-Base.zero(::Type{GenericAffExpr{C,V}}) where {C,V} = GenericAffExpr{C,V}(V[],C[],zero(C))
-Base.one(::Type{GenericAffExpr{C,V}}) where { C,V} = GenericAffExpr{C,V}(V[],C[], one(C))
+function GenericAffExpr(constant::V, kv::AbstractArray{Pair{K,V}}) where {K,V}
+    result = GenericAffExpr{V,K}(constant, new_ordered_dict(K, V, kv))
```
Why doing `result =` and not just `return` ?

blegat commented on src/affexpr.jl:
```diff
@@ -100,51 +170,39 @@ end
+    other_nozeros = dropzeros(other)
+    # Note: This depends on equality of OrderedDicts ignoring order.
+    # This is the current behavior, but it seems questionable.
+    return isequal(aff_nozeros.terms, other_nozeros.terms) && aff.constant == other.constant
```
Why not call `isequal(aff_nozeros, other_nozeros)` ?

blegat commented on src/parseexpr.jl:
```diff
@@ -22,36 +22,34 @@ function parseIdxSet(arg::Expr)
-        x = copy(x)
-        scale!(x.coeffs, c)
-        x.constant *= c
+        x = map_coefficients(x, coef -> c*coef)
```
This is equivalent but less readable than `x = c * x`

blegat commented on src/affexpr.jl:
```diff
@@ -42,48 +102,58 @@ function value(a::GenericAffExpr{T, V}, map::Function) where {T, V}
 end
 
-# Add an affine expression to an existing affine expression
+# TODO: Consider renaming. append! is an unusual name to add two expressions.
```
Or `addterm!` and `addterms!` which is like `MOI.addvariable!` and `MOI.addvariables!`

mlubin commented:
I refactored `push!`, `append!`, and `addtoexpr` into:

```julia
help?> JuMP.add_to_expression!
  add_to_expression!(expression, terms...)

  Updates expression in place to expression + (*)(terms...). This is typically much more efficient
  than expression += (*)(terms...). For example, add_to_expression!(expression, a, b) produces the
  same result as expression += a*b, and add_to_expression!(expression, a) produces the same result as
  expression += a.

  Only a few methods are defined, mostly for internal use, and only for the cases when (1) they can be
  implemented efficiently and (2) expression is capable of storing the result. For example,
  add_to_expression!(::AffExpr, ::VariableRef, ::VariableRef) is not defined because a GenericAffExpr
  cannot store the product of two variables.
```
and
```julia
help?> JuMP.destructive_add!
  destructive_add!(expression, terms...)

  Returns the result of expression + (*)(terms...) and possibly destroys expression (i.e., reusing its
  storage for the result, if possible). For example result = destructive_add!(expression, 2, x) sets
  result to the value of expression + 2*x. expression should not be referred to after this call.

  Note: Discarding the result of destructive_add! suggests an improper usage.
```

Big win for clarity.

I'll merge this after getting the diff coverage sufficiently close to 100%.

mlubin commented:
I'm happy with the coverage now.
@blegat @joehuchette please take a look.

blegat commented on src/affexpr.jl:
```diff
@@ -15,23 +15,84 @@
-    vars::Vector{VarType}
-    coeffs::Vector{CoefType}
     constant::CoefType
+    terms::OrderedDict{VarType,CoefType}
```
Having `terms` as the first field would be more consistent with `MOI.ScalarAffineFunction`

mlubin replied:
I'm aware of the inconsistency with `MOI.ScalarAffineFunction`, but I like the `AffExpr(constant, pairs...)` constructor, and this ordering is consistent with that.

blegat commented on src/affexpr.jl:
```diff
@@ -15,23 +15,84 @@
+
+GenericAffExpr{C, V}() where {C, V} = zero(GenericAffExpr{C, V})
+
+function map_coefficients!(a::GenericAffExpr, f::Function)
```
It would be more consistent with `Base.map` to have `f` as a first parameter

mlubin replied:
Good point, done. I renamed `map_coefficients!` to `map_coefficients_inplace!`, because `map!` is not an in-place operation (it only stores the output in a given container).

blegat commented on src/operators.jl:
```diff
@@ -18,63 +18,107 @@
-Base.:-(lhs::AbstractVariableRef) = GenericAffExpr([lhs],[-1.0],0.0)
-Base.:*(lhs::AbstractJuMPScalar) = lhs # make this more generic so extensions don't have to define unary multiplication for our macros
+Base.:-(lhs::AbstractVariableRef) = GenericAffExpr(0.0, lhs => -1.0)
+Base.:*(lhs::AbstractVariableRef) = lhs # make this more generic so extensions don't have to define unary multiplication for our macros
```
By replacing `AbstractJuMPScalar` with `AbstractVariableRef`, it won't work anymore for `AffExpr` and `QuadExpr`. This should be covered by the tests :/

mlubin replied:
Done and added tests.

blegat commented on src/operators.jl:
```diff
@@ -83,120 +127,152 @@ function Base.:^(lhs::Union{AbstractVariableRef,GenericAffExpr}, rhs::Integer)
-# Don't fall back on AbstractVariableRef*GenericAffExpr to preserve lhs/rhs consistency (appears in printing)
-function Base.:*(lhs::GenericAffExpr{T,V}, rhs::V) where {T,V}
-    n = length(lhs.vars)
+function Base.:+(lhs::GenericAffExpr{C,V}, rhs::V) where {C, V <: AbstractVariableRef}
```
I don't think we need `V <: AbstractVariableRef` since this is already implied by the fact that `V` is the second parameter of `GenericAffExpr{C,V}`

mlubin replied:
We specify `V <: AbstractVariableRef` in lots of places in the overloads for `AbstractVariableRef` above, why shouldn't it be here?

blegat replied:
As a second thought, it's probably better to leave it. We might change the method at some point and it might be needed then so it's safer to be explicit

blegat commented on src/operators.jl:
```diff
@@ -83,120 +127,152 @@ function Base.:^(lhs::Union{AbstractVariableRef,GenericAffExpr}, rhs::Integer)
 Base.:-(lhs::GenericQuadExpr, rhs::Number) = (+)(-rhs,lhs)
 Base.:*(lhs::GenericQuadExpr, rhs::Number) = (*)(rhs,lhs)
-Base.:/(lhs::GenericQuadExpr, rhs::Number) = (*)(inv(rhs),lhs)
+Base.:/(lhs::GenericQuadExpr, rhs::Number) = (*)(1.0/rhs,lhs)
```
Why replacing `inv` ? Isn't it more specific ?

mlubin replied:
Don't remember why I changed this.

blegat commented on src/operators.jl:
```diff
@@ -210,12 +286,12 @@ _sizehint_expr!(q, n) = nothing
 
 # TODO: specialize sum for Dict and JuMPArray of JuMP objects?
-Base.sum(j::Array{<:AbstractVariableRef}) = GenericAffExpr(vec(j), ones(length(j)), 0.0)
+Base.sum(vars::Array{<:AbstractVariableRef}) = GenericAffExpr(0.0, [v => 1.0 for v in vars])
```
Isn't more inference-friendly to use `map` instead of a generator ? I always try to avoid them because `eltype(v => -1.0 for v in vars)` is `Any`

mlubin replied:
I added a test with `@inferred` and this passes.
